api.lisp

```diff
@@ -0,0 +1,77 @@
+;;;; -*- Mode: Lisp; indent-tabs-mode: nil -*-
+
+(in-package :split-sequence)
+
+(declaim (inline list-long-enough-p check-bounds check-tests))
+
+(defun list-long-enough-p (list length)
+  (or (zerop length)
+      (not (null (nthcdr (1- length) list)))))
+
+(defun check-bounds (sequence start end)
+  (progn
+    (check-type start unsigned-byte "a non-negative integer")
+    (check-type end (or null unsigned-byte) "a non-negative integer or NIL")
+    (typecase sequence
+      (list
```

Comment:
This branch is unnecessary because the function list-long-enough-p doesn't do anything the default branch does: NTHCDR is just as O(n) as LENGTH on lists.

Reply:
Sure, both are linear, but they differ in the number of cons cells they will walk.

The idea here is that if we want a subsequence of ten cells from a million-cell-long list, this `NTHCDR` call should walk only ten cells while `LENGTH` will walk the full million.

Reply:
I'd keep the extra branch.  `O(min(end, len(list)))` is better than `O(len(list))`, and a library like this (providing a basic, widely used utility) is probably a good place to spend an extra line or two of code to improve performance.